rfc/001-composable-pipelines/README.md

```diff
@@ -0,0 +1,393 @@
+```
+
+### Synopsis
+
+The core of this proposal introduces pipeline operators as function with input
+and output types. The following overview introduces the pipeline building
+blocks used throughout this proposal.
+
+#### I/O
+
+| Operator | Input Type | Output Type | Description
+| -------- | ---------- | ----------- | ------------------------------------------
+| `from`   | `Void`     | `Bytes`     | Loads bytes from a source as a side effect
+| `to`     | `Bytes`    | `Void`      | Writes bytes to a sink as a side effect
+| `read`   | `Bytes`    | `Arrow`     | Parses a specific format into Arrow
+| `write`  | `Arrow`    | `Bytes`     | Prints input in a specific format
+| `pull`   | `Void`     | `Arrow`     | Loads Arrow from a remote source
+| `push`   | `Arrow`    | `Void`      | Stores Arrow in a remote source
+
+#### Compute
+
+The following operators have a fixed input and output type of `Arrow`. We can
+group them into *filter* operators that do not change the schema, and *reshape*
+operators where execution yields a new schema.
+
+##### Filter
+
+Filter operators have input and output type `Arrow`.
+
+| Operator | Non-Blocking | Description
+| -------- |:------------:| ------------------------------------------
+| `head`   | ✅           | Retrieve first N records
+| `tail`   | ❌           | Retrieve last N records
+| `where`  | ✅           | Filter with expression
+| `sort`   | ❌           | Sort according to field
+| `uniq`   | ✅           | Produce deduplicated output
+
+##### Reshape
+
+Reshape operators have input and output type `Arrow`.
+
+| Operator    | Non-Blocking | Description
+| ----------- |:------------:| ------------------------------------------
+| `put`       | ✅           | Select a set of columns (projection)
+| `drop`      | ✅           | Remove a set of columns (projection)
+| `replace`   | ✅           | Replaces field described by extractors with values
+| `extend`    | ✅           | Adds new fields with (initially fixed) values
+| `rename`    | ✅           | Renames schema meta data (type & fields)
+| `summarize` | ❌/✅[^1]    | Aggregate group and compute a summary function
+
+[^1]: Depending on the aggregate function.
+
+##### Matcher Plugin
+
+| Operator | Input Type | Output Type | Description
+| -------- | ---------- | ----------- | ------------------------------------------
+| `build`  | `Arrow`    | `Matcher`   | Creates a matcher from a data stream
+| `match`  | `Arrow`    | `Arrow`     | Matches input based on its state
+
+#### Commands
+
+The following table illustrate how existing VAST commands map to pipeline
+execution.
+
+| Command Synatx                    | Pipeline Syntax
+| --------------------------------- | -------------------------------------------
+| `vast infer`                      | `vast exec 'read | infer`
+| `vast import FORMAT EXPR`         | `vast exec 'read FORMAT | where EXPR | push'`
+| `vast export FORMAT EXPR`         | `vast spawn 'where EXPR | write FORMAT'`
```

Comment:
For symmetry reasons, I sort of expect the to be a pull, but it is also redundant because there's only one possible data sink in that pipeline, which is the local process.

Reply:
We decided that we are going to remove "remoting" from this proposal and create a dedicated one for it. I'll do that soon.